aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -1628,12 +1628,24 @@ impl AptosVM {
             ChangeSetConfigs::unlimited_at_gas_feature_version(self.gas_feature_version);
 
         match write_set_payload {
-            WriteSetPayload::Direct(change_set) => VMChangeSet::try_from_storage_change_set(
-                change_set.clone(),
-                &change_set_configs,
-                resolver.is_delayed_field_optimization_capable(),
-            )
-            .map_err(|e| e.into_vm_status()),
+            WriteSetPayload::Direct(change_set) => {
+                // this transaction is never delayed field capable.
+                // it is required it requires restarting execution afterwards,
+                // which allows it to be used as last transaction in delayed_field_enabled context.
+                let change = VMChangeSet::try_from_storage_change_set_with_delayed_field_optimization_disabled(
```

Comment:
So this populates resource change set in `VMChangeSet` with groups as well (which is wrong), but after all still works when we convert it later to `TransactionOutput` because we merge everything together. This seems a bit fragile to me...

Reply:
even if resource groups flag is set, this will have non-separated resource group - basically there will be a single resource representing the whole group - i.e. the materialized version.

after this transactions, reads from view are not correct (as accessing resource group elements will not see correct things), but that is why the function is named so long and there are warnings around.

it is fragile, but I don't have a better alternative. do you have a better suggestion?

Reply:
Sadly I am not sure we can do anything better😂 constructing a direct materialized change set does not seem better and requires quite a bit of changes.